File: src/cs/Demos/SolidEdge.OpenSave/InteropProxy.cs
using log4net;
using log4net.Config;
using SolidEdge.Common;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;

namespace SolidEdge.OpenSave
{
    public class InteropProxy : MarshalByRefObject
    {
        private static ILog _log = LogManager.GetLogger(typeof(InteropProxy));
        private OpenSaveSettings _openSaveSettings;

        public void DoOpenSave(string filePath, OpenSaveSettings openSaveSettings)
        {
            _openSaveSettings = openSaveSettings;

            // Register with OLE to handle concurrency issues on the current thread.
            OleMessageFilter.Register();

            try
            {
                SolidEdgeFramework.Application application = ApplicationHelper.Connect(true);
                SolidEdgeFramework.Documents documents = null;
                SolidEdgeFramework.SolidEdgeDocument document = null;

                application.Visible = _openSaveSettings.Application.Visible;

                if (_openSaveSettings.Application.DisableAddins == true)
                {
                    DisableAddins(application);
                }

                // Disable (most) prompts.
                application.DisplayAlerts = false;

                // Get a reference to the documents collection.
                documents = application.Documents;

                // Close any documents that may be left open for whatever reason.
                documents.Close();

                // Open the file.
                document = (SolidEdgeFramework.SolidEdgeDocument)documents.Open(filePath);

                application.DoIdle();

                if (document != null)
                {
                    // Environment specific routines.
                    if (document is SolidEdgeAssembly.AssemblyDocument)
                    {
                        DoOpenSave((SolidEdgeAssembly.AssemblyDocument)document, _openSaveSettings.Assembly);
                    }
                    else if (document is SolidEdgeDraft.DraftDocument)
                    {
                        DoOpenSave((SolidEdgeDraft.DraftDocument)document, _openSaveSettings.Draft);
                    }
                    else if (document is SolidEdgePart.PartDocument)
                    {
                        DoOpenSave((SolidEdgePart.PartDocument)document, _openSaveSettings.Part);
                    }
                    else if (document is SolidEdgePart.SheetMetalDocument)
                    {
                        DoOpenSave((SolidEdgePart.SheetMetalDocument)document, _openSaveSettings.SheetMetal);
                    }
                    else if (document is SolidEdgePart.WeldmentDocument)
                    {
                        DoOpenSave((SolidEdgePart.WeldmentDocument)document, _openSaveSettings.Weldment);
                    }

                    // Save document.
                    document.Save();

                    // Close document.
                    document.Close();

                    application.DoIdle();
                }
            }
            finally
            {
                OleMessageFilter.Unregister();
            }
        }

        private void DoOpenSave(SolidEdgeAssembly.AssemblyDocument assemblyDocument, AssemblySettings assemblySettings)
        {
        }

        private void DoOpenSave(SolidEdgeDraft.DraftDocument draftDocument, DraftSettings draftSettings)
        {
            SolidEdgeDraft.Sections sections = null;
            SolidEdgeDraft.Section section = null;
            SolidEdgeDraft.SectionSheets sectionSheets = null;
            SolidEdgeDraft.Sheet sheet = null;
            SolidEdgeDraft.DrawingViews drawingViews = null;
            SolidEdgeDraft.DrawingView drawingView = null;

            if (draftSettings.UpdateDrawingViews)
            {
                sections = draftDocument.Sections;
                section = sections.WorkingSection;

                sectionSheets = section.Sheets;

                for (int i = 1; i <= sectionSheets.Count; i++)
                {
                    sheet = sectionSheets.Item(i);
                    drawingViews = sheet.DrawingViews;

                    for (int j = 1; j <= drawingViews.Count; j++)
                    {
                        drawingView = drawingViews.Item(j);
                        drawingView.Update();
                    }
                }
            }
        }

        private void DoOpenSave(SolidEdgePart.PartDocument partDocument, PartSettings partSettings)
        {
        }

        private void DoOpenSave(SolidEdgePart.SheetMetalDocument sheetMetalDocument, SheetMetalSettings sheetMetalSettings)
        {
        }

        private void DoOpenSave(SolidEdgePart.WeldmentDocument weldmentDocument, WeldmentSettings weldmentSettings)
        {
        }

        private void DisableAddins(SolidEdgeFramework.Application application)
        {
            SolidEdgeFramework.AddIns addins = null;
            SolidEdgeFramework.AddIn addin = null;

            addins = application.AddIns;

            for (int i = 1; i <= addins.Count; i++)
            {
                addin = addins.Item(i);
                addin.Connect = false;
            }
        }
    }
}

File: vb/SolidEdge/OpenSave/OpenSave/InteropProxy.vb
Imports log4net
Imports log4net.Config
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Runtime.InteropServices
Imports System.Text

Public Class InteropProxy
	Inherits MarshalByRefObject

	Private Shared _log As ILog = LogManager.GetLogger(GetType(InteropProxy))
	Private _openSaveSettings As OpenSaveSettings

	Public Sub DoOpenSave(ByVal filePath As String, ByVal openSaveSettings As OpenSaveSettings)
		_openSaveSettings = openSaveSettings

		' Register with OLE to handle concurrency issues on the current thread.
		SolidEdgeCommunity.OleMessageFilter.Register()

		Try
			Dim application As SolidEdgeFramework.Application = SolidEdgeCommunity.SolidEdgeUtils.Connect(True)
			Dim documents As SolidEdgeFramework.Documents = Nothing
			Dim document As SolidEdgeFramework.SolidEdgeDocument = Nothing

			application.Visible = _openSaveSettings.Application.Visible

			If _openSaveSettings.Application.DisableAddins = True Then
				DisableAddins(application)
			End If

			' Disable (most) prompts.
			application.DisplayAlerts = False

			' Get a reference to the documents collection.
			documents = application.Documents

			' Close any documents that may be left open for whatever reason.
			documents.Close()

			' Open the file.
			document = DirectCast(documents.Open(filePath), SolidEdgeFramework.SolidEdgeDocument)

			application.DoIdle()

			If document IsNot Nothing Then
				' Environment specific routines.
				If TypeOf document Is SolidEdgeAssembly.AssemblyDocument Then
					DoOpenSave(DirectCast(document, SolidEdgeAssembly.AssemblyDocument), _openSaveSettings.Assembly)
				ElseIf TypeOf document Is SolidEdgeDraft.DraftDocument Then
					DoOpenSave(DirectCast(document, SolidEdgeDraft.DraftDocument), _openSaveSettings.Draft)
				ElseIf TypeOf document Is SolidEdgePart.PartDocument Then
					DoOpenSave(DirectCast(document, SolidEdgePart.PartDocument), _openSaveSettings.Part)
				ElseIf TypeOf document Is SolidEdgePart.SheetMetalDocument Then
					DoOpenSave(DirectCast(document, SolidEdgePart.SheetMetalDocument), _openSaveSettings.SheetMetal)
				ElseIf TypeOf document Is SolidEdgePart.WeldmentDocument Then
					DoOpenSave(DirectCast(document, SolidEdgePart.WeldmentDocument), _openSaveSettings.Weldment)
				End If

				' Save document.
				document.Save()

				' Close document.
				document.Close()

				application.DoIdle()
			End If
		Finally
			SolidEdgeCommunity.OleMessageFilter.Unregister()
		End Try
	End Sub

	Private Sub DoOpenSave(ByVal assemblyDocument As SolidEdgeAssembly.AssemblyDocument, ByVal assemblySettings As AssemblySettings)
	End Sub

	Private Sub DoOpenSave(ByVal draftDocument As SolidEdgeDraft.DraftDocument, ByVal draftSettings As DraftSettings)
		Dim sections As SolidEdgeDraft.Sections = Nothing
		Dim section As SolidEdgeDraft.Section = Nothing
		Dim sectionSheets As SolidEdgeDraft.SectionSheets = Nothing
		Dim sheet As SolidEdgeDraft.Sheet = Nothing
		Dim drawingViews As SolidEdgeDraft.DrawingViews = Nothing
		Dim drawingView As SolidEdgeDraft.DrawingView = Nothing

		If draftSettings.UpdateDrawingViews Then
			sections = draftDocument.Sections
			section = sections.WorkingSection

			sectionSheets = section.Sheets

			For i As Integer = 1 To sectionSheets.Count
				sheet = sectionSheets.Item(i)
				drawingViews = sheet.DrawingViews

				For j As Integer = 1 To drawingViews.Count
					drawingView = drawingViews.Item(j)
					drawingView.Update()
				Next j
			Next i
		End If
	End Sub

	Private Sub DoOpenSave(ByVal partDocument As SolidEdgePart.PartDocument, ByVal partSettings As PartSettings)
	End Sub

	Private Sub DoOpenSave(ByVal sheetMetalDocument As SolidEdgePart.SheetMetalDocument, ByVal sheetMetalSettings As SheetMetalSettings)
	End Sub

	Private Sub DoOpenSave(ByVal weldmentDocument As SolidEdgePart.WeldmentDocument, ByVal weldmentSettings As WeldmentSettings)
	End Sub

	Private Sub DisableAddins(ByVal application As SolidEdgeFramework.Application)
		Dim addins As SolidEdgeFramework.AddIns = Nothing
		Dim addin As SolidEdgeFramework.AddIn = Nothing

		addins = application.AddIns

		For i As Integer = 1 To addins.Count
			addin = addins.Item(i)
			addin.Connect = False
		Next i
	End Sub
End Class
